File: kmail_password_recovery/recovery.cpp
#include <iostream>

#include <qstring.h>


QString decrypt(const QString &aStr)
{
    QString result;
    for (unsigned int i = 0; i < aStr.length(); i++)
        result += (aStr[i].unicode() <= 0x21 ) ? aStr[i] :
		                            QChar(0x1001F - aStr[i].unicode());
    return result;
}


int main(int argc, char** argv)
{
    if (argc != 2)
    {
        std::cerr << "usage: ./recovery <password>" << std::endl;
	return 1;
    }

    QString pwd = QString::fromUtf8(argv[1]);
    std::cout << decrypt(pwd) << std::endl;
    return 0;
}

File: kmail_password_recovery/recovery.py
#!/usr/bin/env python

import os.path
import sys
from ConfigParser import ConfigParser, NoOptionError


def decrypt(data):
    """Python equivalent of kmail's KMAccount::encryptStr"""
    result = list()
    for char in data:
        if ord(char) <= 0x21:
            result.append(char)
        else:
            result.append(unichr(0x1001F - ord(char)))
    return ''.join(result)


def main(args=sys.argv[1:]):
    if len(sys.argv) != 2:
        exec_file = os.path.basename(sys.argv[0])
        print >> sys.stderr, 'Usage: %s <config file>' % exec_file
        return 1

    config = ConfigParser()
    config.read(args[0])
    for section in config.sections():
        if not section.startswith('Transport '):
            continue
        name = config.get(section, 'name')
        try:
            pwd = decrypt(config.get(section, 'pass').decode('utf-8'))
        except NoOptionError:
            pwd = str()
        print '%s:%s' % (name, pwd)
    
    return 0


if __name__ == '__main__':
    sys.exit(main())
